Can SVT-AV1 generate xcode project ?

Hi SVT experts, 

I want to debug the SVT-AV1 code by using the xcode in MacOS, so could you please tell me how to generate the xcode project for SVT-AV1 ?

@joelsole , Thank you for your reply. 
when I run Build/XCode/build.sh, It can generate the xcode project, but when I build, it has many errors, such as: 
Undefined symbols for architecture x86_64:
  "_Log2f_SSE2", referenced from:
      _encode_transform in libCOMMON_CODEC.a(EbTransforms.o)
      _encode_inv_transform in libCOMMON_CODEC.a(EbTransforms.o)
      _product_full_loop in libCOMMON_CODEC.a(EbFullLoop.o)
      _cu_full_distortion_fast_tu_mode_r in libCOMMON_CODEC.a(EbFullLoop.o)
      _inject_inter_candidates in libCOMMON_CODEC.a(EbModeDecision.o)
      _ProductGenerateMdCandidatesCu in libCOMMON_CODEC.a(EbModeDecision.o)
      _md_scan_all_blks in libCOMMON_CODEC.a(EbUtility.o)
      ...
  "_RunEmms", referenced from:
      _get_proj_subspace_c in libCOMMON_CODEC.a(EbRestorationPick.o)
      _search_wiener_seg in libCOMMON_CODEC.a(EbRestorationPick.o)
      _search_wiener in libCOMMON_CODEC.a(EbRestorationPick.o)
      _compute_score in libCOMMON_CODEC.a(EbRestorationPick.o)
      _search_selfguided_restoration in libCOMMON_CODEC.a(EbRestorationPick.o)
      _search_wiener_finish in libCOMMON_CODEC.a(EbRestorationPick.o)
      _get_proj_subspace_avx2 in libCOMMON_ASM_AVX2.a(EbRestorationPick_AVX2.o)
      ...
  "_aom_dc_128_predictor_16x16_sse2", referenced from:
      _setup_rtcd_internal in EbEncHandle.o
  "_aom_dc_128_predictor_4x4_sse2", referenced from:
      _setup_rtcd_internal in EbEncHandle.o
  "_aom_dc_128_predictor_8x8_sse2", referenced from:
      _setup_rtcd_internal in EbEncHandle.o
  "_aom_dc_left_predictor_16x16_sse2", referenced from:
      _setup_rtcd_internal in EbEncHandle.o
  "_aom_dc_left_predictor_4x4_sse2", referenced from:
      _setup_rtcd_internal in EbEncHandle.o
  "_aom_dc_left_predictor_8x8_sse2", referenced from:
      _setup_rtcd_internal in EbEncHandle.o

--- a/.github/workflows/macos.yml
+++ b/.github/workflows/macos.yml
@@ -1,0 +1,44 @@
+name: macOS
+on:
+  - push
+  - pull_request
+env:
+  build_type: Release
+  native_deps: nasm yasm cmake ccache
+
+jobs:
+  build:
+    runs-on: macos-10.15
+    env:
+      CC: clang
+      CXX: clang++
+      generator: Xcode
+    steps:
+    - name: Install dependencies
+      run: |
+        brew update
+        brew install yasm ccache python || true
+        echo "::set-env name=PATH::/usr/local/opt/ccache/libexec:$PATH"
+        clang -v
+        clang++ -v
+    - name: Print system info
+      run: |
+        sysctl -n machdep.cpu.brand_string
+        sysctl machdep.cpu.features
+        sysctl machdep.cpu.leaf7_features
+    - uses: actions/checkout@v2
+    - name: Cache ccache files
+      uses: actions/cache@v1.1.2
+      with:
+        path: ~/.ccache
+        key: ${{ runner.os }}-svt-av1-clang-${{ hashFiles('**/*.c') }}
+        restore-keys: |
+          ${{ runner.os }}-svt-av1-clang-
+    - name: Run CMake
+      run: cmake -S "$GITHUB_WORKSPACE" -B Build -G"${generator:=Unix Makefiles}" -DCMAKE_BUILD_TYPE="${build_type:=Release}" $CMAKE_EFLAGS
+
+    - name: Run Build
+      run: cmake --build Build --parallel 4 --config "${build_type:=Release}"
+
+    - name: Install
+      run: sudo cmake --build Build --config "${build_type:=Release}" --target install

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -154,6 +154,50 @@ if(NOT HAS_AVX512)
     add_definitions(-DNON_AVX512_SUPPORT)
 endif()
 
+# ASM compiler macro
+macro(ASM_COMPILE_TO_TARGET target)
+    if(CMAKE_GENERATOR STREQUAL "Xcode")
+        if(CMAKE_BUILD_TYPE MATCHES Release)
+            set(ASM_FLAGS "${CMAKE_ASM_NASM_FLAGS_RELEASE}")
+        elseif(CMAKE_BUILD_TYPE MATCHES Debug)
+            set(ASM_FLAGS "${CMAKE_ASM_NASM_FLAGS_DEBUG}")
+        elseif(CMAKE_BUILD_TYPE MATCHES MinSizeRel)
+            set(ASM_FLAGS "${CMAKE_ASM_NASM_FLAGS_MINSIZEREL}")
+        elseif(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
+            set(ASM_FLAGS "${CMAKE_ASM_NASM_FLAGS_RELWITHDEBINFO}")
+        endif()
+
+        get_property(inc_dirs DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY INCLUDE_DIRECTORIES)
+        foreach(inc_dir ${inc_dirs})
+            set(ASM_FLAGS "${ASM_FLAGS} -I${inc_dir}")
+        endforeach()
+
+        string(REGEX REPLACE " " ";" ASM_FLAGS "${ASM_FLAGS} ${CMAKE_ASM_NASM_FLAGS}")
+
+        foreach(asm_file ${ARGN})
+            get_filename_component(filename "${asm_file}" NAME_WE)
+            set(SIMD_OBJ "${CMAKE_CURRENT_BINARY_DIR}/${filename}${CMAKE_C_OUTPUT_EXTENSION}")
+            message(STATUS "Pregenerating ASM_NASM file ${filename}${CMAKE_C_OUTPUT_EXTENSION}")
+            execute_process(COMMAND ${CMAKE_ASM_NASM_COMPILER}
+                    -f${CMAKE_ASM_NASM_OBJECT_FORMAT}
+                    -I${CMAKE_CURRENT_SOURCE_DIR}
+                    -o${SIMD_OBJ}
+                    ${ASM_FLAGS}
+                    ${CMAKE_CURRENT_SOURCE_DIR}/${asm_file}
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR})
+            list(APPEND SIMD_OBJS "${SIMD_OBJ}")
+        endforeach()
+        #target_sources(${target} PUBLIC ${SIMD_OBJS})
+        target_sources(common_lib INTERFACE ${SIMD_OBJS})
+    else()
+        # For every other generator
+        foreach(asm_file ${ARGN})
+            get_filename_component(filename "${asm_file}" ABSOLUTE)
+            target_sources(${target} PUBLIC ${filename})
+        endforeach()
+    endif()
+endmacro()
+
 # Add Subdirectories
 add_subdirectory(Source/Lib/Common)
 if(BUILD_ENC)

--- a/Source/Lib/Common/ASM_SSE2/CMakeLists.txt
+++ b/Source/Lib/Common/ASM_SSE2/CMakeLists.txt
@@ -21,15 +21,15 @@ test_apply_compiler_flags(${flags_to_test})
 
 file(GLOB all_files
     "*.h"
-    "*.asm"
     "*.c")
 
-file(GLOB asm_includes
-    "x86_abi_support.asm"
-    "x64Macro.asm"
-    "x64inc.asm"
-    "x86inc.asm")
-
-list(REMOVE_ITEM all_files ${asm_includes})
+set(asm_files
+    intrapred_sse2.asm
+    EbPictureOperators_SSE2.asm
+    highbd_intrapred_sse2_.asm
+    aom_subpixel_8t_sse2.asm
+    subtract_sse2.asm)
 
 add_library(COMMON_ASM_SSE2 OBJECT ${all_files})
+
+asm_compile_to_target(COMMON_ASM_SSE2 ${asm_files})

--- a/Source/Lib/Common/ASM_SSE2/x64inc.asm
+++ b/Source/Lib/Common/ASM_SSE2/x64inc.asm
@@ -3,6 +3,42 @@
 ; SPDX - License - Identifier: BSD - 2 - Clause - Patent
 ;
 
+%undef WIN64
+%undef UNIX64
+%ifidn __OUTPUT_FORMAT__,win32
+    %define WIN64  1
+%elifidn __OUTPUT_FORMAT__,win64
+    %define WIN64  1
+%elifidn __OUTPUT_FORMAT__,x64
+    %define WIN64  1
+%else
+    %define UNIX64 1
+%endif
+
+%undef FORMAT_ELF
+%ifidn __OUTPUT_FORMAT__,elf
+    %define FORMAT_ELF 1
+%elifidn __OUTPUT_FORMAT__,elf32
+    %define FORMAT_ELF 1
+%elifidn __OUTPUT_FORMAT__,elf64
+    %define FORMAT_ELF 1
+%endif
+
+%undef FORMAT_MACHO
+%ifidn __OUTPUT_FORMAT__,macho32
+     %define FORMAT_MACHO 1
+%elifidn __OUTPUT_FORMAT__,macho64
+     %define FORMAT_MACHO 1
+%endif
+
+%ifdef FORMAT_ELF
+    %undef PREFIX
+%elif WIN64
+    %undef PREFIX
+%else
+    %define PREFIX
+%endif
+
 %ifdef PREFIX
     %define mangle(x) _ %+ x
 %else

--- a/Source/Lib/Common/ASM_SSSE3/CMakeLists.txt
+++ b/Source/Lib/Common/ASM_SSSE3/CMakeLists.txt
@@ -25,11 +25,11 @@ test_apply_compiler_flags(${flags_to_test})
 
 file(GLOB all_files
     "*.h"
-    "*.asm"
     "*.c")
 
-file(GLOB asm_includes "../ASM_SSE2/x86_abi_support.asm")
-
-#list(REMOVE_ITEM all_files ${asm_includes})
+set(asm_files
+    aom_subpixel_bilinear_ssse3.asm)
 
 add_library(COMMON_ASM_SSSE3 OBJECT ${all_files})
+
+asm_compile_to_target(COMMON_ASM_SSSE3 ${asm_files})

--- a/Source/Lib/Common/CMakeLists.txt
+++ b/Source/Lib/Common/CMakeLists.txt
@@ -24,6 +24,8 @@ else ()
     set (X86 FALSE)
 endif ()
 
+add_library(common_lib INTERFACE)
+
 if(X86)
 add_subdirectory(Codec)
 add_subdirectory(C_DEFAULT)

--- a/Source/Lib/Decoder/CMakeLists.txt
+++ b/Source/Lib/Decoder/CMakeLists.txt
@@ -10,113 +10,89 @@ set(DEC_VERSION_MINOR 8)
 set(DEC_VERSION_PATCH 1)
 set(DEC_VERSION ${DEC_VERSION_MAJOR}.${DEC_VERSION_MINOR}.${DEC_VERSION_PATCH})
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
-    set (X86 TRUE)
-else ()
-    set (X86 FALSE)
-endif ()
+    set(X86 TRUE)
+else()
+    set(X86 FALSE)
+endif()
+
+if(UNIX)
+    if(NOT APPLE)
+        find_library(M_LIB name m)
+        if(M_LIB)
+            list(APPEND PLATFORM_LIBS m)
+        endif()
+    endif()
+    set(LIBS_PRIVATE "-lpthread -lm")
+endif()
+
+if(COVERAGE)
+    if(CMAKE_C_COMPILER_ID MATCHES "GNU")
+        set(LIBS_PRIVATE "-lgcov ${LIBS_PRIVATE}")
+    else(CMAKE_C_COMPILER_ID MATCHES "Clang")
+        set(LIBS_PRIVATE "--coverage ${LIBS_PRIVATE}")
+    endif()
+endif()
+
+include_directories(${PROJECT_BINARY_DIR}/Source/Lib/Common/Codec/
+    ${PROJECT_SOURCE_DIR}/Source/API/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Decoder/Codec/
+    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
+link_directories(${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Decoder/Codec/
+    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
 
 if(X86)
-# Include Decoder Subdirectories
-include_directories(${PROJECT_BINARY_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/API/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Decoder/Codec/
-    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
+    # Include Decoder Subdirectories
+    include_directories(${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/)
 
-link_directories(
-     ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Decoder/Codec/
-    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
+    link_directories(${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/)
+endif()
+
+# Required for cmake to be able to tell Xcode how to link all of the object files
+if(CMAKE_GENERATOR STREQUAL "Xcode")
+    file(WRITE ${PROJECT_BINARY_DIR}/dummy.c "")
+endif()
 
 file(GLOB all_files
     "Codec/*.h"
     "Codec/*.c")
 
-if(UNIX)
-    if(NOT APPLE)
-        find_library(M_LIB name m)
-        if(M_LIB)
-            list(APPEND PLATFORM_LIBS m)
-        endif()
-    endif()
-    set(LIBS_PRIVATE "-lpthread -lm")
-endif()
-
-if(COVERAGE)
-    if(CMAKE_C_COMPILER_ID MATCHES "GNU")
-        set(LIBS_PRIVATE "-lgcov ${LIBS_PRIVATE}")
-    else(CMAKE_C_COMPILER_ID MATCHES "Clang")
-        set(LIBS_PRIVATE "--coverage ${LIBS_PRIVATE}")
-    endif()
+get_target_property(common_lib_source common_lib INTERFACE_SOURCES)
+if(common_lib_source)
+    list(APPEND all_files ${common_lib_source} ${PROJECT_BINARY_DIR}/dummy.c)
 endif()
 
 # Decoder Lib Source Files
-add_library(SvtAv1Dec
-    ${all_files}
-    $<TARGET_OBJECTS:COMMON_CODEC>
-    $<TARGET_OBJECTS:FASTFEAT>
-    $<TARGET_OBJECTS:COMMON_C_DEFAULT>
-    $<TARGET_OBJECTS:COMMON_ASM_SSE2>
-    $<TARGET_OBJECTS:COMMON_ASM_SSSE3>
-    $<TARGET_OBJECTS:COMMON_ASM_SSE4_1>
-    $<TARGET_OBJECTS:COMMON_ASM_AVX2>
-    $<TARGET_OBJECTS:COMMON_ASM_AVX512>)
-else ()
-# Include Decoder Subdirectories
-include_directories(${PROJECT_BINARY_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/API/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Decoder/Codec/
-    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
-
-link_directories(
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Decoder/Codec/
-    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
-
-file(GLOB all_files
-    "Codec/*.h"
-    "Codec/*.c")
-
-if(UNIX)
-    if(NOT APPLE)
-        find_library(M_LIB name m)
-        if(M_LIB)
-            list(APPEND PLATFORM_LIBS m)
-        endif()
-    endif()
-    set(LIBS_PRIVATE "-lpthread -lm")
+if(X86)
+    add_library(SvtAv1Dec
+        ${all_files}
+        $<TARGET_OBJECTS:COMMON_CODEC>
+        $<TARGET_OBJECTS:FASTFEAT>
+        $<TARGET_OBJECTS:COMMON_C_DEFAULT>
+        $<TARGET_OBJECTS:COMMON_ASM_SSE2>
+        $<TARGET_OBJECTS:COMMON_ASM_SSSE3>
+        $<TARGET_OBJECTS:COMMON_ASM_SSE4_1>
+        $<TARGET_OBJECTS:COMMON_ASM_AVX2>
+        $<TARGET_OBJECTS:COMMON_ASM_AVX512>)
+else()
+    add_library(SvtAv1Dec
+        ${all_files}
+        $<TARGET_OBJECTS:COMMON_CODEC>
+        $<TARGET_OBJECTS:FASTFEAT>
+        $<TARGET_OBJECTS:COMMON_C_DEFAULT>)
 endif()
 
-if(COVERAGE)
-    if(CMAKE_C_COMPILER_ID MATCHES "GNU")
-        set(LIBS_PRIVATE "-lgcov ${LIBS_PRIVATE}")
-    else(CMAKE_C_COMPILER_ID MATCHES "Clang")
-        set(LIBS_PRIVATE "--coverage ${LIBS_PRIVATE}")
-    endif()
-endif()
-
-# Decoder Lib Source Files
-add_library(SvtAv1Dec
-    ${all_files}
-    $<TARGET_OBJECTS:COMMON_CODEC>
-    $<TARGET_OBJECTS:FASTFEAT>
-    $<TARGET_OBJECTS:COMMON_C_DEFAULT>)
-endif ()
 set_target_properties(SvtAv1Dec PROPERTIES VERSION ${DEC_VERSION})
 set_target_properties(SvtAv1Dec PROPERTIES SOVERSION ${DEC_VERSION_MAJOR})
 add_dependencies(SvtAv1Dec EbVersionHeaderGen)

--- a/Source/Lib/Encoder/ASM_SSE2/CMakeLists.txt
+++ b/Source/Lib/Encoder/ASM_SSE2/CMakeLists.txt
@@ -23,13 +23,12 @@ test_apply_compiler_flags(${flags_to_test})
 
 file(GLOB all_files
     "*.h"
-    "*.asm"
     "*.c")
 
-file(GLOB asm_includes
-    "x86_abi_support.asm")
-
-#list(REMOVE_ITEM all_files)
-
+set(asm_files
+    highbd_variance_impl_sse2.asm
+    x64RegisterUtil.asm)
 
 add_library(ENCODER_ASM_SSE2 OBJECT ${all_files})
+
+asm_compile_to_target(ENCODER_ASM_SSE2 ${asm_files})

--- a/Source/Lib/Encoder/ASM_SSSE3/CMakeLists.txt
+++ b/Source/Lib/Encoder/ASM_SSSE3/CMakeLists.txt
@@ -25,10 +25,11 @@ test_apply_compiler_flags(${flags_to_test})
 
 file(GLOB all_files
     "*.h"
-    "*.asm"
     "*.c")
 
-file(GLOB asm_includes "x86_abi_support.asm")
-list(REMOVE_ITEM all_files ${asm_includes})
+#set(asm_files
+#    x86_abi_support.asm)
 
 add_library(ENCODER_ASM_SSSE3 OBJECT ${all_files})
+
+#asm_compile_to_target(ENCODER_ASM_SSSE3 ${asm_files})

--- a/Source/Lib/Encoder/CMakeLists.txt
+++ b/Source/Lib/Encoder/CMakeLists.txt
@@ -10,160 +10,120 @@ set(ENC_VERSION_MINOR 8)
 set(ENC_VERSION_PATCH 1)
 set(ENC_VERSION ${ENC_VERSION_MAJOR}.${ENC_VERSION_MINOR}.${ENC_VERSION_PATCH})
 
-if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
-    set (X86 TRUE)
-else ()
-    set (X86 FALSE)
-endif ()
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
+    set(X86 TRUE)
+else()
+    set(X86 FALSE)
+endif()
 
+if(UNIX)
+    if(NOT APPLE)
+        find_library(M_LIB name m)
+        if(M_LIB)
+            list(APPEND PLATFORM_LIBS m)
+        endif()
+    endif()
+    set(LIBS_PRIVATE "-lpthread -lm")
+endif()
+
+if(COVERAGE)
+    if(CMAKE_C_COMPILER_ID MATCHES "GNU")
+        set(LIBS_PRIVATE "-lgcov ${LIBS_PRIVATE}")
+    else(CMAKE_C_COMPILER_ID MATCHES "Clang")
+        set(LIBS_PRIVATE "--coverage ${LIBS_PRIVATE}")
+    endif()
+endif()
+
+include_directories(${PROJECT_SOURCE_DIR}/Source/API/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/C_DEFAULT/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Codec/
+    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
+link_directories(${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/C_DEFAULT/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Codec/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Globals/)
+
+add_subdirectory(C_DEFAULT)
+add_subdirectory(Codec)
+add_subdirectory(Globals)
 if(X86)
-# Include Encoder Subdirectories
-include_directories(${PROJECT_SOURCE_DIR}/Source/API/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSSE3/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE4_1/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX512/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Codec/
-    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
+    # Include Encoder Subdirectories
+    include_directories(${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSSE3/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE4_1/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX512/)
 
-link_directories(
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSSE3/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE4_1/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX2/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX512/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Globals/
-)
+    link_directories(
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSSE3/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_SSE4_1/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX2/
+        ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/ASM_AVX512/)
+    add_subdirectory(ASM_SSE2)
+    add_subdirectory(ASM_SSSE3)
+    add_subdirectory(ASM_SSE4_1)
+    add_subdirectory(ASM_AVX2)
+    add_subdirectory(ASM_AVX512)
+endif()
 
-add_subdirectory(C_DEFAULT)
-add_subdirectory(ASM_SSE2)
-add_subdirectory(ASM_SSSE3)
-add_subdirectory(ASM_SSE4_1)
-add_subdirectory(ASM_AVX2)
-add_subdirectory(ASM_AVX512)
-add_subdirectory(Codec)
-add_subdirectory(Globals)
-
-#file(GLOB all_files
-#    "Codec/*.h"
-#    "Codec/*.c")
+# Required for cmake to be able to tell Xcode how to link all of the object files
+if(CMAKE_GENERATOR STREQUAL "Xcode")
+    file(WRITE ${PROJECT_BINARY_DIR}/dummy.c "")
+endif()
 
 file(GLOB all_files
     "*.h"
     "*.c")
 
-if(UNIX)
-    if(NOT APPLE)
-        find_library(M_LIB name m)
-        if(M_LIB)
-            list(APPEND PLATFORM_LIBS m)
-        endif()
-    endif()
-    set(LIBS_PRIVATE "-lpthread -lm")
-endif()
-
-if(COVERAGE)
-    if(CMAKE_C_COMPILER_ID MATCHES "GNU")
-        set(LIBS_PRIVATE "-lgcov ${LIBS_PRIVATE}")
-    else(CMAKE_C_COMPILER_ID MATCHES "Clang")
-        set(LIBS_PRIVATE "--coverage ${LIBS_PRIVATE}")
-    endif()
+get_target_property(common_lib_source common_lib INTERFACE_SOURCES)
+if(common_lib_source)
+    list(APPEND all_files ${common_lib_source} ${PROJECT_BINARY_DIR}/dummy.c)
 endif()
 
 # Encoder Lib Source Files
-add_library(SvtAv1Enc
-    ${all_files}
-    $<TARGET_OBJECTS:COMMON_CODEC>
-    $<TARGET_OBJECTS:FASTFEAT>
-    $<TARGET_OBJECTS:COMMON_C_DEFAULT>
-    $<TARGET_OBJECTS:COMMON_ASM_SSE2>
-    $<TARGET_OBJECTS:COMMON_ASM_SSSE3>
-    $<TARGET_OBJECTS:COMMON_ASM_SSE4_1>
-    $<TARGET_OBJECTS:COMMON_ASM_AVX2>
-    $<TARGET_OBJECTS:COMMON_ASM_AVX512>
-    $<TARGET_OBJECTS:ENCODER_GLOBALS>
-    $<TARGET_OBJECTS:ENCODER_CODEC>
-    $<TARGET_OBJECTS:ENCODER_C_DEFAULT>
-    $<TARGET_OBJECTS:ENCODER_ASM_SSE2>
-    $<TARGET_OBJECTS:ENCODER_ASM_SSSE3>
-    $<TARGET_OBJECTS:ENCODER_ASM_SSE4_1>
-    $<TARGET_OBJECTS:ENCODER_ASM_AVX2>
-    $<TARGET_OBJECTS:ENCODER_ASM_AVX512>)
-else ()
-# Include Encoder Subdirectories
-include_directories(${PROJECT_SOURCE_DIR}/Source/API/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Codec/
-    ${PROJECT_SOURCE_DIR}/third_party/fastfeat/)
-
-link_directories(
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/C_DEFAULT/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Codec/
-    ${PROJECT_SOURCE_DIR}/Source/Lib/Encoder/Globals/
-)
-
-add_subdirectory(C_DEFAULT)
-add_subdirectory(Codec)
-add_subdirectory(Globals)
-
-#file(GLOB all_files
-#    "Codec/*.h"
-#    "Codec/*.c")
-
-file(GLOB all_files
-    "*.h"
-    "*.c")
-
-if(UNIX)
-    if(NOT APPLE)
-        find_library(M_LIB name m)
-        if(M_LIB)
-            list(APPEND PLATFORM_LIBS m)
-        endif()
-    endif()
-    set(LIBS_PRIVATE "-lpthread -lm")
+if(X86)
+    add_library(SvtAv1Enc
+        ${all_files}
+        $<TARGET_OBJECTS:COMMON_CODEC>
+        $<TARGET_OBJECTS:FASTFEAT>
+        $<TARGET_OBJECTS:COMMON_C_DEFAULT>
+        $<TARGET_OBJECTS:COMMON_ASM_SSE2>
+        $<TARGET_OBJECTS:COMMON_ASM_SSSE3>
+        $<TARGET_OBJECTS:COMMON_ASM_SSE4_1>
+        $<TARGET_OBJECTS:COMMON_ASM_AVX2>
+        $<TARGET_OBJECTS:COMMON_ASM_AVX512>
+        $<TARGET_OBJECTS:ENCODER_GLOBALS>
+        $<TARGET_OBJECTS:ENCODER_CODEC>
+        $<TARGET_OBJECTS:ENCODER_C_DEFAULT>
+        $<TARGET_OBJECTS:ENCODER_ASM_SSE2>
+        $<TARGET_OBJECTS:ENCODER_ASM_SSSE3>
+        $<TARGET_OBJECTS:ENCODER_ASM_SSE4_1>
+        $<TARGET_OBJECTS:ENCODER_ASM_AVX2>
+        $<TARGET_OBJECTS:ENCODER_ASM_AVX512>)
+else()
+    add_library(SvtAv1Enc
+        ${all_files}
+        $<TARGET_OBJECTS:COMMON_CODEC>
+        $<TARGET_OBJECTS:FASTFEAT>
+        $<TARGET_OBJECTS:COMMON_C_DEFAULT>
+        $<TARGET_OBJECTS:ENCODER_GLOBALS>
+        $<TARGET_OBJECTS:ENCODER_CODEC>
+        $<TARGET_OBJECTS:ENCODER_C_DEFAULT>)
 endif()
-
-if(COVERAGE)
-    if(CMAKE_C_COMPILER_ID MATCHES "GNU")
-        set(LIBS_PRIVATE "-lgcov ${LIBS_PRIVATE}")
-    else(CMAKE_C_COMPILER_ID MATCHES "Clang")
-        set(LIBS_PRIVATE "--coverage ${LIBS_PRIVATE}")
-    endif()
-endif()
-
-# Encoder Lib Source Files
-add_library(SvtAv1Enc
-    ${all_files}
-    $<TARGET_OBJECTS:COMMON_CODEC>
-    $<TARGET_OBJECTS:FASTFEAT>
-    $<TARGET_OBJECTS:COMMON_C_DEFAULT>
-    $<TARGET_OBJECTS:ENCODER_GLOBALS>
-    $<TARGET_OBJECTS:ENCODER_CODEC>
-    $<TARGET_OBJECTS:ENCODER_C_DEFAULT>)
-endif ()
 
 set_target_properties(SvtAv1Enc PROPERTIES VERSION ${ENC_VERSION})
 set_target_properties(SvtAv1Enc PROPERTIES SOVERSION ${ENC_VERSION_MAJOR})

